Support CSC/BSR/BSC inputs to unary zero-preserving functions

pearu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #78173
* #79062

In addition, enable testing masked reductions in sparse compressed consistency check.

pearu committed:
Support CSC/BSR/BSC inputs to torch.abs

In addition, enable testing masked reductions in sparse compressed consistency check.

[ghstack-poisoned]

pearu committed:
Update on "Support CSC/BSR/BSC inputs to torch.abs"


In addition, enable testing masked reductions in sparse compressed consistency check.

[ghstack-poisoned]

pearu committed:
Update on "Support CSC/BSR/BSC inputs to torch.abs"


In addition, enable testing masked reductions in sparse compressed consistency check.

[ghstack-poisoned]

pearu committed:
Update on "Support CSC/BSR/BSC inputs to torch.abs"


In addition, enable testing masked reductions in sparse compressed consistency check.

[ghstack-poisoned]

pearu committed:
Update on "Support CSC/BSR/BSC inputs to torch.abs"


In addition, enable testing masked reductions in sparse compressed consistency check.

[ghstack-poisoned]

cpuhrsch commented on test/test_sparse_csr.py:
```diff
@@ -425,6 +428,81 @@ def test_copy_errors(self, layout, device, dtype):
+                    continue
+            expected = op(sample.input, **sample.kwargs)
+            assert torch.is_tensor(expected)
+            blocksize = tuple(map(self._smallest_divisor, sample.input.shape[-2:]))
```
Can you add a comment that describes why the blocksize was chosen this?

pearu replied:
Added comment:
```
# Use smallest non-trivial blocksize for the given input shape:
```

cpuhrsch commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -10496,6 +10537,8 @@ def error_inputs_mean(op_info, device, **kwargs):
                        # RuntimeError: "nonzero_cuda" not implemented for 'ComplexHalf'
                        DecorateInfo(unittest.expectedFailure, 'TestSparseCSR', 'test_sparse_csr_consistency',
                                     dtypes=(torch.chalf,)),
+                       DecorateInfo(unittest.expectedFailure, 'TestSparseCompressed', 'test_consistency',
```
It seems almost easier to add ComplexHalf support to nonzero than to add this DecorateInfo to all these tests. From a brief look at it, perhaps it's enough to just [extend the types that we're generating code for](https://github.com/pytorch/pytorch/blob/35d4a805ebc3b6eca1bafb2d332dffa8d0c1fc54/aten/src/ATen/native/cuda/Nonzero.cu#L121-L124)?

cpuhrsch replied:
There's a `AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4` and it seems to be used for pretty much exactly this case: https://github.com/pytorch/pytorch/blob/f7ee308dfb6c1f28421562eca7685a7043c9f845/aten/src/ATen/native/cpu/CopyKernel.cpp#L84-L86

I'd still send that change in a separate PR in case it touches some unexpected piece, but adding `kComplexHalf` to nonzero seems worthwhile based on how many tests you need to skip here and given that it's needed for construction of COO Tensors in general.

pearu replied:
Resolved via https://github.com/pytorch/pytorch/pull/79062

pearu committed:
Update on "Support CSC/BSR/BSC inputs to torch.abs"


In addition, enable testing masked reductions in sparse compressed consistency check.

[ghstack-poisoned]

pearu committed:
Update on "Support CSC/BSR/BSC inputs to torch.abs"


In addition, enable testing masked reductions in sparse compressed consistency check.

[ghstack-poisoned]

pearu committed:
Update on "Support CSC/BSR/BSC inputs to torch.abs"


In addition, enable testing masked reductions in sparse compressed consistency check.

[ghstack-poisoned]

pearu committed:
Update on "Support CSC/BSR/BSC inputs to torch.abs"


In addition, enable testing masked reductions in sparse compressed consistency check.

[ghstack-poisoned]

pearu commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -10304,6 +10339,9 @@ def error_inputs_mean(op_info, device, **kwargs):
                    supports_fwgrad_bwgrad=True,
                    assert_autodiffed=True,
                    supports_sparse_csr=True,
+                   supports_sparse_csc=True,
```
Done.

pearu committed:
Update on "Support CSC/BSR/BSC inputs to unary zero-preserving functions"


In addition, enable testing masked reductions in sparse compressed consistency check.

[ghstack-poisoned]

pearu committed:
Update on "Support CSC/BSR/BSC inputs to unary zero-preserving functions"


In addition, enable testing masked reductions in sparse compressed consistency check.

[ghstack-poisoned]

pearu commented:
@pytorchbot merge this please

pearu commented:
@pytorchbot help

pearu commented:
@pytorchbot merge -g

pearu committed:
Update on "Support CSC/BSR/BSC inputs to unary zero-preserving functions"


In addition, enable testing masked reductions in sparse compressed consistency check.

[ghstack-poisoned]

pearu commented:
@pytorchbot merge -a

pearu commented:
@pytorchbot merge -a
